src/coniql/strawberry_schema.py

```diff
@@ -0,0 +1,263 @@
+
+
+@strawberry.type
+class ChannelTime(TypeChannelTime):
+    """
+    Timestamp indicating when a value was last updated.
+    Inherit fields from TypeChannelTime.
+    """
+
+    @strawberry.field
+    async def datetime(root: TypeChannelTime) -> datetime.datetime:
+        """The timestamp as a datetime object"""
+        return datetime.datetime.fromtimestamp(root.seconds)
+
+
+async def resolver_value(root: DeferredChannel) -> Optional[TypeChannelValue]:
+    channel = await root.get_channel()
+    return channel.get_value()
+
+
+async def resolver_time(root: DeferredChannel) -> Optional[TypeChannelTime]:
+    channel = await root.get_channel()
+    return channel.get_time()
+
+
+async def resolver_status(root: DeferredChannel) -> Optional[TypeChannelStatus]:
+    channel = await root.get_channel()
+    return channel.get_status()
+
+
+async def resolver_display(root: DeferredChannel) -> Optional[ChannelDisplay]:
+    channel = await root.get_channel()
+    return channel.get_display()
+
+
+@strawberry.type
+class Channel:
+    """
+    A single value with associated time, status and metadata. These values
+    can be Null so that in a subscription they are only updated on change
+    """
+
+    # ID that uniquely defines this Channel, normally a PV
+    id: strawberry.ID
+    # The current value of this channel
+    value: Optional[ChannelValue] = strawberry.field(resolver=resolver_value)
+    # When was the value last updated
+    time: Optional[ChannelTime] = strawberry.field(resolver=resolver_time)
+    # Status of the connection, whether is is mutable, and alarm info
+    status: Optional[ChannelStatus] = strawberry.field(resolver=resolver_status)
+    # How should the Channel be displayed
+    display: Optional[ChannelDisplay] = strawberry.field(resolver=resolver_display)
+
+
+def get_channel(id: strawberry.ID, timeout: float = 5.0) -> TypeChannel:
+    return GetChannel(id, timeout, store_global)
+
+
+@strawberry.type
+class Query:
+    # Get the current value of a Channel
+    getChannel: Channel = strawberry.field(resolver=get_channel)
+
+
+async def subscribe_channel(id: strawberry.ID) -> AsyncGenerator[TypeChannel, None]:
+    """Subscribe to changes in top level fields of Channel,
+    if they haven't changed they will be Null"""
+    store: PluginStore = store_global
+    plugin, channel_id = store.plugin_config_id(id)
+    # Remove the transport prefix from the read pv
+    pv = store.transport_pv(id)[1]
+    async for channel in plugin.subscribe_channel(pv):
+        yield SubscribeChannel(channel_id, channel)
+
+
+@strawberry.type
+class Subscription:
+    """Tell mypy to ignore this line as it complains 'expression
+    has type "T", variable has type "Channel"'. Possibly a Strawberry issue
+    as strawberry.subscription is a partial function"""
+
+    subscribeChannel: Optional[Channel] = strawberry.subscription(
+        resolver=subscribe_channel
+    )  # type: ignore
```

Comment:
More for my own interest than an actual problem, but why is this an `Optional`? I don't understand the need to override the type hinting (and why we suppress mypy's checks) here.

Reply:
Not obvious so maybe I should add a comment but when I was testing with multiple machine-status screens that were all unsubscribing and then re-subscribing I noticed that during the disconnect you could get a lot of exceptions as the Channel was not returned as the websocket had been closed. Adding `Optional` to this fixes these errors as it doesn't complain if the Channel isn't returned.

mypy reports the type of the `strawberry.subscription(resolver=...)` as type "T". Looking in the Strawberry code I found that this is actually a partial function. I couldn't find a way to get a type that mypy was happy with so had to add the ignore.